[Dynamo] Implement MATCH_CLASS opcode for structural pattern matching

GSumanth109 opened the pull request:
Fixes #171274 

### Summary
This PR implements the `MATCH_CLASS` opcode in Dynamo, enabling support for Python 3.10+ structural pattern matching (`match`/`case`) with classes.

Previously, using `case ClassName():` would trigger an `Unsupported: missing: MATCH_CLASS` error. This PR adds the missing logic to `symbolic_convert.py` to correctly handle class matching, attribute extraction, and fall-through behavior.

**Key Implementation Details:**
1.  **Opcode Support:** Implemented `MATCH_CLASS`, which pops `names`, `cls_type`, and `subject` from the stack.
2.  **Variable Realization:** Added `.realize()` calls for arguments to ensure `isinstance` checks work correctly against `LazyVariableTracker`.
3.  **Safety Logic (Tensor vs Class):** Added an explicit check for `TensorVariable` vs `UserDefinedClassVariable` mismatches.
    * *Context:* Dynamo's internal `builtins.isinstance` handler raises an `AssertionError` when checking if a `Tensor` is an instance of a user class.
    * *Fix:* We explicitly catch this case and return `False` (no match) instead of crashing the compiler.
4.  **Stack Management:** Correctly restores the `subject` to the stack in the failure case. This is required for Python to proceed to the next `case` block (fixing a potential `IndexError`).

### Test Plan
I verified the fix using a comprehensive test script covering:
1.  **The Bug:** Matching a Tensor against a User Class (should fail gracefully, not crash).
2.  **Success Case:** Successfully matching a class instance.
3.  **Attribute Extraction:** extracting values like `case Point(x=val):`.
4.  **Built-in Types:** Matching against `int`, `float`, etc.

**Verification Script:**
```python
import unittest
import torch
import torch._dynamo
from dataclasses import dataclass

@dataclass
class Point:
    x: int
    y: int

class TestDynamoMatchClass(unittest.TestCase):
    def setUp(self):
        torch._dynamo.reset()

    def test_tensor_vs_class_mismatch(self):
        """The original bug: A Tensor should NOT match a User Class."""
        @torch.compile(backend="eager")
        def fn(x):
            match x:
                case Point(): return "matched_point"
                case _: return "mismatch_fallback"

        x = torch.randn(2)
        # Should NOT crash, should return fallback
        self.assertEqual(fn(x), "mismatch_fallback")

    def test_class_match_success(self):
        """Verify we can actually match a class correctly."""
        @torch.compile(backend="eager")
        def fn(obj):
            match obj:
                case Point(): return "matched_point"
                case _: return "failed"

        p = Point(1, 2)
        self.assertEqual(fn(p), "matched_point")

    def test_attribute_extraction(self):
        """Verify we can extract attributes: case Point(x=val):"""
        @torch.compile(backend="eager")
        def fn(obj):
            match obj:
                case Point(x=val): return val
                case _: return -1

        p = Point(x=10, y=20)
        self.assertEqual(fn(p), 10)

    def test_builtin_type_match(self):
        """Verify standard types work: case int():"""
        @torch.compile(backend="eager")
        def fn(x):
            match x:
                case int(): return "is_int"
                case _: return "other"

        self.assertEqual(fn(5), "is_int")

if __name__ == "__main__":
    unittest.main()
```
**Output:**
```
Ran 4 tests in 0.184s

OK
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

GSumanth109 committed:
[Dynamo] Implement MATCH_CLASS opcode for structural pattern matching

Lucaskabela reviewed (changes requested):
Please fix failing CI signals.

In addition, the first thing we need (preferably before writing impl details) is unit tests to check our code behavior

Since this is supporting a new control structure which may commonly be used, you should add extensive testing to our code base

Please search for where tracing test for existing control structures lives, and add tests for MATCH there covering MATCH in a variety of uses and with different types

Once done, we can closer inspect the implementation details

GSumanth109 committed:
Fix Pyre errors, lint issues, and add unit test for MATCH_CLASS

GSumanth109 commented:
Thanks for the review, @Lucaskabela! I think I've finally wrangled the linter and the type checker. 

I've pushed the fixes for the points we discussed:
- **Pyre/Types:** Switched to using lists for `call_function` arguments, which satisfied the type checker.
- **Linter:** Fixed the `B904` error by explicitly adding `from None` to the exception raise.
- **Safety:** Added the `if match_result is None` check so it properly falls back to a graph break instead of crashing if `isinstance` fails.
- **Tests:** Added the new test case to cover both the Dataclass match and the failure/fallback scenario.

Ran the tests and linter locally and everything looks green now. Ready for another look when you have a chance!

GSumanth109 commented:
@pytorchbot label "topic: not user facing"

guilhermeleobas commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3699,6 +3699,70 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+            self.push(names)
+            raise Unsupported(
+                "MATCH_CLASS: cannot statically resolve isinstance"
+            ) from None
```
This doesn't look correct to me. You're pushing the attributes on `NotImplementedError`

guilhermeleobas commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3699,6 +3699,70 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+        if hasattr(cls_type, "realize"):
+            cls_type = cls_type.realize()
+        if hasattr(subject, "realize"):
+            subject = subject.realize()
```
It should be safe to call `.realize()` on all three variable trackers without checking `hasattr(obj, "realize")`.

GSumanth109 committed:
Address review: cleanup impl and port CPython test_patma.py cases

GSumanth109 committed:
Address review: use cleaner implementation and add CPython tests

GSumanth109 commented:
Thanks for the pointers @guilhermeleobas !

1. **Implementation:** You were right regarding the cleanup—I removed the explicit `Tensor` check. `BuiltinVariable(isinstance)` returns `None` for Tensors, which correctly raises `Unsupported` and triggers a safe graph break/fallback. The implementation is now fully cleaned up (no `hasattr` checks, no manual stack pushes).

2. **Tests:** I ported the relevant class-pattern tests directly from CPython's `test_patma.py` as requested. The new suite covers:
    * `test_patma_177` (Positional & Keyword args)
    * `test_patma_183` (Guards on classes)
    * `test_patma_200` (Custom `__match_args__`)
    * `test_patma_146` (Built-in types like `int`)
    * `test_patma_111` (Class attributes)

guilhermeleobas commented:
@GSumanth109, can you copy the file to [`test/dynamo/cpython/3_13/test_patma.py`](https://github.com/pytorch/pytorch/tree/main/test/dynamo/cpython/3_13)? We have some infra in place to run the CPython tests under dynamo:

1. copy `[test_patma.py](https://raw.githubusercontent.com/python/cpython/refs/tags/v3.13.5/Lib/test/test_patma.py)` to `test/dynamo/cpython/3_13/`
2. patch the test file to update the test base class and the function executed inside `__main__`. See an example [here](https://github.com/pytorch/pytorch/blob/main/test/dynamo/cpython/3_13/test_bool.diff)
3. Create a patch using git diff
4. run the test file with python 3.13 and dynamo with `PYTORCH_TEST_WITH_DYNAMO=1 python test/dynamo/cpython/3_13/test_patma.py`
5. Add the failures to `test/dynamo_expected_failures/CPython313-test_patma-TESTCLASS-TESTNAME`

You might want to do this in a separated PR as this will be quite verbose.

GSumanth109 commented:
@guilhermeleobas Thanks for the detailed instructions on the CPython integration!

Integrating the full CPython test suite seems like a great step for ensuring long-term robustness, and I'd be happy to take that on properly.

Given that the infrastructure setup and expected failures list will be quite verbose, would it be alright if we merge this PR with the current targeted tests (which already cover the `MATCH_CLASS` cases)?

I can then immediately open a follow-up PR dedicated to setting up `test/dynamo/cpython/3_13/test_patma.py` as you suggested. Does that plan sound good to you?

Lucaskabela commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3699,6 +3699,48 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
         else:
             self.push(ConstantVariable.create(False))
 
+    def MATCH_CLASS(self, inst):
```
nit: let's add type hints to this API

GSumanth109 committed:
Add backend tests and fix types (linted)

GSumanth109 committed:
Merge remote-tracking branch 'upstream/main' into fix/dynamo-match-class
git push origin fix/dynamo-match-class
ase enter a commit message to explain why this merge is necessary,

GSumanth109 commented:
@Lucaskabela Thanks for the feedback! I've pushed the updates:

1. **Backend Verification**: I added `test/dynamo/test_match_backends.py` to explicitly test the feature against **Eager**, **AOT**, and **Inductor** backends (including the `nn.Module` case).
   * *Rationale*: I kept this in a separate file to cleanly distinguish *backend stability testing* (does it survive compilation?) from the *feature logic testing* (does it handle attributes/nesting?) in `test_functions.py`.
2. **Type Hints**: Updated the `MATCH_CLASS` signature.

@guilhermeleobas As discussed, I am working on it and I'll try to open the follow-up PR for the full CPython `test_patma.py` integration immediately after this lands.

GSumanth109 commented on test/dynamo/test_match_backends.py:
```diff
@@ -0,0 +1,70 @@
+                    return torch.zeros(1)
+
+        x = torch.randn(4)
+        self.assertTrue(same(fn(x), x * 2))
```
@guilhermeleobas Thanks for checking! I included `aot_eager` and `inductor` specifically because @Lucaskabela requested coverage for them in his review ("hitting eager, aot_eager, and inductor backends").

I am happy to keep them for robustness (to ensure no backend-specific lowering issues), but let me know if you strictly prefer removing them.

GSumanth109 committed:
Merge upstream/main and resolve conflict

guilhermeleobas commented:
Send it as a separated PR, please

Lucaskabela commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,48 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
         else:
             self.push(ConstantVariable.create(False))
 
+    def MATCH_CLASS(self, inst: Instruction):
```
nit: add return
```
def MATCH_CLASS(self, inst: Instruction) -> None:
```

Lucaskabela commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,48 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+        except NotImplementedError:
+            raise Unsupported(
+                "MATCH_CLASS: cannot statically resolve isinstance"
+            ) from None
```
we don't need `from None` here, do we?

Lucaskabela reviewed (changes requested):
Few more thoughts but looks close to ready

Lucaskabela commented:
> Send it as a separated PR, please

+1 to this

GSumanth109 committed:
Address reviewer feedback: Refactor tests and fix nits

GSumanth109 committed:
Address reviewer feedback: Refactor tests and fix nits

Lucaskabela reviewed (approved):
Thanks for responding to the feedback - generally LGTM :)

guilhermeleobas commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,46 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+        except AssertionError:
+            match_result = ConstantVariable.create(False)
+        except Exception:
+            raise
```
Let the exception propagate instead of capturing it


```suggestion
    match_result = BuiltinVariable(builtins.ininstance).call_function(self, [subject, cls_type], {})
```

guilhermeleobas commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,46 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+        try:
+            is_match = match_result.as_python_constant()
+        except NotImplementedError:
+            raise Unsupported("MATCH_CLASS: cannot statically resolve isinstance")  # noqa: B904
```
Looking at the impl. of `isinstance` in dynamo, the return variable is a constant variable, so, it should be safe to just look its value.


```suggestion
        if istype(match_result, ConstantVariable):
            is_match = match_result.value
        else:
            unimplemented(...)
```

guilhermeleobas reviewed (changes requested):
Hi @GSumanth109, I gave the PR another look. Thanks again for working on this, we are on the right track.

guilhermeleobas commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,46 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+            raise
+
+        if match_result is None:
+            raise Unsupported("MATCH_CLASS: isinstance returned None")
```
See an example of `unimplemented` being used here:

https://github.com/pytorch/pytorch/blob/df7916747debf3eb7135670b65489b88e0296e35/torch/_dynamo/symbolic_convert.py#L577-L586

GSumanth109 committed:
Address review: fix MATCH_CLASS logic, imports, and error handling

GSumanth109 committed:
Address review: clean up MATCH_CLASS logic and remove dead code

williamwen42 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,41 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+        subject = self.pop().realize()
+
+        try:
+            match_result = self.call_function(
```
Prefer to use a VT's call_function over InstructionTranslatorBase's call_function

williamwen42 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,41 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+            match_result = self.call_function(
+                BuiltinVariable(isinstance), [subject, cls_type], {}
+            )
+        except AssertionError:
```
What kind of errors could happen here?

williamwen42 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -3715,6 +3715,41 @@ def MATCH_MAPPING(self, inst: Instruction) -> None:
+                context="dynamic isinstance",
+                explanation="MATCH_CLASS: cannot statically resolve isinstance",
+                hints=[],
+            )
```
According to https://docs.python.org/3/library/dis.html, in 3.10, we also pushed a boolean value indicating success or failure.
